pubsub.go

```diff
@@ -130,6 +150,9 @@ type PubSubRouter interface {
 	// Leave notifies the router that we are no longer interested in a topic.
 	// It is invoked after the unsubscription announcement.
 	Leave(topic string)
+	// Bootstrap gets the bootstrapping discovery parameters from the router.
+	// It also sets the mechanism for retrieving new discovery requests for the topic.
+	Bootstrap() []discovery.Option
```

Comment:
Do routers really need to be able to configure discovery options? IMO, this is more of a global option. That is, that is, "discovery" is just a background service that:

1. Tells the network we're interested in a topic.
2. Connects to more peers when asked.

Reply:
Given our conversation regarding moving the complexity from the router level to the validator and application level perhaps the discovery options (like Limit) could be passed in on a per-topic basis. This could be done by making another register function similar to "RegisterTopicValidator", by adding fields to Subscription and utilizing a SubOpt, or something in between.

Reply:
This relates to https://github.com/libp2p/go-libp2p-pubsub/pull/184#discussion_r285807815. It's one of the reasons I believe the router should be the one asking for more peers.

Basically, it's up to routers to form the pubsub network. They may even have alternative/preferred discovery mechanisms.

Reply:
> Basically, it's up to routers to form the pubsub network. They may even have alternative/preferred discovery mechanisms.

Maybe... under the premises of how I wrote #171 I agreed with you. However, after our conversations with @raulk and @vyzo about moving complexity from new routers into per-topic validators the landscape looks different. Maybe floodsub/gossipsub/episub cares about the discovery mechanisms, but maybe lww-persistence/DAG-persistence/validator that only propagate every third message/etc. cares about the discovery mechanisms.

If we're prepared to go back to "all interesting logic is in the routers" then I'm happy to put discovery there. However, given the addition of networking logic from the application layer (e.g. via stateful validators as we're planning on using for persistence) we might be better off with something like a TopicConfig that ends up containing all the smarts so that the discovery strategy isn't tied to floodsub/gossipsub/etc. but to a struct/function that implements that logic.

Reply:
This looks like a pretty useless method as it has empty implementations, so let's remove it.

Reply:
As indicated above it would be great to be able to say "persistent channels only need to wait for 1-2 peers during bootstrapping instead of 12". Originally I was going to do this via a new router (as in #171), which means that the implementations here would be empty and there would not.

Given that, per our discussion, I've now rewritten #171 to use stateful validators + external protocols this doesn't look super useful. **However**, I'd like to replace this with a PubSub function for per-topic configurations (similar in nature to how topic validators, which we're also using to influence message propagation, are per-topic).

Reply:
Plan is to punt on this for this PR. While it hurts that bootstrapping before publishing can be unnecessarily slow (e.g. waiting for the 10s timeout to find 100 peers when we know we only need 5), having bootstrapping for subscribe be asynchronous takes away the most painful part of the experience.